util-images/network/netperfbenchmark/pkg/util/util.go

```diff
@@ -0,0 +1,178 @@
+package util
+
+import (
+	"errors"
+	"math"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"k8s.io/klog"
+)
+
+// Protocols supported
+const (
+	ProtocolTCP  = "TCP"
+	ProtocolUDP  = "UDP"
+	ProtocolHTTP = "HTTP"
+)
+
+// Number of metrics for each protocol
+const (
+	tcpMtrCnt   = 2
+	udpMtrCnt   = 11
+	httpMetrCnt = 12
+)
+
+// Iperf results vary from protocol,required for proper parsing
+const (
+	zeroFmtTCP         = "0.0"
+	fractionZeroFmtTCP = ".0"
+	zeroFmtUDP         = "0.00"
+	fractionZeroFmtUDP = ".00"
+)
+
+// Function to be applied for each metric for aggregation
+var (
+	iperfUDPFn = []string{"", "", "", "Sum", "Sum", "Sum", "Sum", "Sum", "Avg", "Avg", "Min", "Max", "Avg", "Sum"}
+	iperfTCPFn = []string{"", "", "", "Sum", "Avg"}
+)
+
+// ParseResult parses the response received for each protocol type
+func ParseResult(protType string, result []string, testDuration string) ([]float64, error) {
+	klog.Info("Parsing response for ", protType)
+	switch protType {
+	case ProtocolTCP:
+		return parseIperfResponse(result, testDuration, ProtocolTCP, tcpMtrCnt, zeroFmtTCP, fractionZeroFmtTCP, iperfTCPFn), nil
+	case ProtocolUDP:
+		return parseIperfResponse(result, testDuration, ProtocolUDP, udpMtrCnt, zeroFmtUDP, fractionZeroFmtUDP, iperfUDPFn), nil
+	case ProtocolHTTP:
+		return parseSiegeResponse(result), nil
+	default:
+		return nil, errors.New("invalid protocol:" + protType)
+	}
+}
+
+func parseIperfResponse(result []string, testDuration string, protocol string, metricCnt int, zeroFmt string, fractionZeroFmt string, operType []string) []float64 {
+	klog.Info("In parseIperf")
+	sumResult := make([]float64, 0, metricCnt)
+	unitReg := regexp.MustCompile(`%|\[\s+|\]\s+|KBytes\s+|KBytes/sec\s*|sec\s+|pps\s*|ms\s+|/|\(|\)\s+`)
+	mulSpaceReg := regexp.MustCompile(`\s+`)
+	hypSpcReg := regexp.MustCompile(`\-\s+`)
+	cnt := 0
+	sessionID := make(map[string]bool)
+	for _, op := range result {
+		klog.Info(op)
+		frmtString := hypSpcReg.ReplaceAllString(mulSpaceReg.ReplaceAllString(unitReg.ReplaceAllString(op, " "), " "), "-")
+		if !strings.Contains(frmtString, zeroFmt+"-"+testDuration+fractionZeroFmt) {
+			continue
+		}
+		klog.Info("Trim info:", frmtString)
+		split := strings.Split(frmtString, " ")
+		//for bug in iperf tcp
+		//if the record is for the complete duration of run
+		if len(split) >= metricCnt+1 && "SUM" != split[1] && split[2] == zeroFmt+"-"+testDuration+fractionZeroFmt {
+			if _, ok := sessionID[split[1]]; ok {
+				continue
+			}
+			for i, v := range split {
+				klog.Info("Split", i, ":", v)
+				if i == 1 {
+					sessionID[v] = true
+					continue
+				}
+				//first index and hte last is ""
+				if i == 0 || i == 2 || i == metricCnt+3 {
+					continue
+				}
+				tmp, err := strconv.ParseFloat(v, 64)
+				if err != nil {
+					klog.Error("Conversion error", err)
+				}
+				if len(sumResult) < metricCnt {
+					sumResult = append(sumResult, tmp)
+				} else {
+					switch operType[i] {
```

Comment:
operType is set to "empty" in multiple places - why isn't this handled?

Reply:
The ones that are set to empty are responses from iperf which are not metrics. Those fields would not even reach the switch statement as those are ignored initially itself .